### PR TITLE
Introduce a flat option to ensure_contiguous_ndarray to switch off flatten for ZFPY codec

### LINKED ISSUE #303: Data array is flattened in numcodecs which reduced the compression ratio that ZFP can provide on multi-dimension arrays
#### Minimal, reproducible code sample, a copy-pastable example if possible

```python
In ensure_contiguous_ndarray:
if arr.flags.c_contiguous or arr.flags.f_contiguous:
        # can flatten without copy
        arr = arr.reshape(-1, order='A')

```

#### Problem description

ZFP is a lossy compressor that expects to get better compression ratio when compressing multi-dimension arrays. Currently in the zfpy.encode, ensure_contiguous_ndarray function flattens the array and feeds to ZFP compressor. In this way, ZFP compressor only can get less expected compression ratio. So I tried to add a switch in ensure_contiguous_ndarray, if the codec is ZFP, I skipped the flattening. However, this only works on C-order array, it doesn't work on F-order array because ZFP decoder only returns C-order array. When the decoder returns C-order multi-dimension array for an original F-order multi-dimension array, ZFP compressor cannot pass the F-order array encode and decode test. 

#### Version and installation information

Please provide the following:

* 0.9.0 of ``numcodecs.__version__``
* Python 3.7, 3.8
* Operating system (Linux/Windows)
* pip install

### PR DESCRIPTION
**Pull request by halehawk**
closes #303 

TODO:

- [x] Unit tests and/or doctests in docstrings
- [ x] `tox -e py39` passes locally
- [x] Docstrings and API docs for any new/modified user-facing classes and functions
- [ x] Changes documented in docs/release.rst
- [x ] `tox -e docs` passes locally
- [x] GitHub Actions CI passes
- [x] Test coverage to 100% (Coveralls passes)

### CODE REVIEW

File: numcodecs/zfpy.py [hunk 1]
```diff
@@ -54,8 +55,16 @@ def __init__(
 
         def encode(self, buf):
 
-            # normalise inputs
-            buf = ensure_contiguous_ndarray(buf)
+            # not flatten c-order array and raise exception for f-order array
+            if not isinstance(buf, np.ndarray):
+                raise TypeError("The zfp codec does not support none numpy arrays."
```
Comment (rabernat) [inline numcodecs/zfpy.py:60]: What other type would you expect here? I thought all Zarr array data would come into numcodecs as numpy arrays?

Comment (halehawk) [reply numcodecs/zfpy.py:60]: I would like to repeat my answers here, I added test_err_encode_list, which caused buf.flags not raise a correct error. So I added check the instance first. Do you have any different recommendation?

**Review by rabernat — APPROVED** [review]
Thanks for your work on this! LGTM!

**Comment by rabernat** [comment]
One final question I have (which doesn't have to hold up this PR) is whether you have tested this branch on a real-world use case and verified that it leads to much better compression, as expected from the discussion in #303.

**Comment by halehawk** [comment]
@rabernat I checked both numcodecs 0.9.1 which has flatten array and my modified numcodecs. The non-flatten array definitely can provide us better compression ratio, for example using precison mode 16 bits, we can get from 2.28 to 4.32 on a 4D array, though the error metric might be not that good, such as: rmse from 0.003979 to 0.05241.

**Comment by halehawk** [comment]
Looks like Peter's group added some meta data about macos-python3.10-zfpy build, but didn't upload the package yet. So I got download problem with this zfpy package.

**Comment by halehawk** [comment]
Looks like zfpy-0.5.5-cp310-macos can be pip install successfully now, this time the failed OSX CI/build (3.10) should be able to pass. How can I restart all PR checks?

**Comment by halehawk** [comment]
Now my PR failed on macos-py310 check again, but the problem was happened during building blosc.cpython-310-darwin.so. @rabernat @jakirkham Do you have any advice on this? Can you ignore this failure and merge my PR now? Thanks!

**Comment by rabernat** [comment]
I refreshed the CI. It looks like the OSX 3.10 build is still failing.

**Comment by jakirkham** [comment]
Regarding CI. Fixing in PR ( https://github.com/zarr-developers/numcodecs/pull/311 )

File: .github/workflows/ci-osx.yaml [hunk 2]
```diff
@@ -27,7 +27,7 @@ jobs:
       - name: Set up env
         shell: "bash -l {0}"
         run: |
-          conda create -n env python==${{matrix.python-version}} wheel pip compilers
+          conda create -n env python=${{matrix.python-version}} wheel pip compilers 'clang>=12.0.1'
```
Comment (jakirkham) [inline .github/workflows/ci-osx.yaml:30]: Let's leave this for that PR to sort out 🙂

Comment (halehawk) [reply .github/workflows/ci-osx.yaml:30]: You want to me revert the modification?

Comment (jakirkham) [reply .github/workflows/ci-osx.yaml:30]: Yes

**Comment by rabernat** [comment]
LGTM, but i would appreciate another maintainer approval.

File: numcodecs/compat.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def ensure_ndarray(buf):
     return arr
 
 
-def ensure_contiguous_ndarray(buf, max_buffer_size=None):
+def ensure_contiguous_ndarray(buf, max_buffer_size=None, flat=True):
```
Comment (jakirkham) [inline numcodecs/compat.py:53]: Preference for using a verb to describe the action taken or not

```suggestion
def ensure_contiguous_ndarray(buf, max_buffer_size=None, flatten=True):
```

File: numcodecs/compat.py [hunk 4]
```diff
@@ -91,8 +91,10 @@ def ensure_contiguous_ndarray(buf, max_buffer_size=None):
 
     # check memory is contiguous, if so flatten
     if arr.flags.c_contiguous or arr.flags.f_contiguous:
-        # can flatten without copy
-        arr = arr.reshape(-1, order='A')
+        # check if flat flag is on or not
+        if flat:
```
Comment (jakirkham) [inline numcodecs/compat.py:95]:
```suggestion
        # check if flatten flag is on or not
        if flatten:
```

**Review by jakirkham — COMMENTED** [review]
Generally LGTM. Had a small suggestion on parameter naming. Once done would be happy to merge 🙂 

Thanks for all the work here @halehawk! 😄

**Review by jakirkham — APPROVED** [review]

**Comment by halehawk** [comment]
Done. Thanks for the feedback.

File: numcodecs/zfpy.py [hunk 5]
```diff
@@ -54,8 +55,16 @@ def __init__(
...
+                                f" Your buffers were {type(buf)}.")
+            if buf.flags.c_contiguous:
+                flat = False
+            else:
+                raise ValueError("The zfp codec does not support F order arrays. "
+                                 f"Your arrays flags were {buf.flags}.")
+            buf = ensure_contiguous_ndarray(buf, flat=flat)
```
Comment (jakirkham) [inline numcodecs/zfpy.py:67]:
```suggestion
            buf = ensure_contiguous_ndarray(buf, flatten=flatten)
```

File: numcodecs/zfpy.py [hunk 6]
```diff
@@ -54,8 +55,16 @@ def __init__(
...
         def encode(self, buf):
 
-            # normalise inputs
-            buf = ensure_contiguous_ndarray(buf)
+            # not flatten c-order array and raise exception for f-order array
+            if not isinstance(buf, np.ndarray):
+                raise TypeError("The zfp codec does not support none numpy arrays."
+                                f" Your buffers were {type(buf)}.")
+            if buf.flags.c_contiguous:
+                flat = False
```
Comment (jakirkham) [inline numcodecs/zfpy.py:63]:
```suggestion
                flatten = False
```

**Review by jakirkham — COMMENTED** [review]
Think this should fix CI

**Comment by jakirkham** [comment]
Thanks @halehawk for the PR and everyone for the reviews! 😄

**Pull request merged**